### ISSUE TITLE
XTDB should work by default on OSX

### ISSUE THREAD

**Issue by armincerf** [xtdb/xtdb#894]
**UPDATE [2024-02-11 - @refset]** per the comment in the thread below, OSX users must currently set the following env vars before JVM startup (e.g. as a prefix to the `clj` command): `XTDB_DISABLE_LIBGCRYPT=true XTDB_ENABLE_BYTEUTILS_SHA1=true`

###  Original report

This is likely caused by something specific to my machine (running OSX 10.15) as it works on my desktop (also running OSX 10.15).

If I have a blank project with `juxt/crux-core {:mvn/version "20.05-1.8.3-alpha"}` in a deps.edn and run clj, typing `(require '[crux.api :as crux])` in the repl causes clojure to crash. A full example is below:

```
Clojure 1.10.1
user=> (require '[crux.api :as crux])
clojure: warning: clojure crashed, killed by SIGABRT.
clojure itself has not crashed, but for transparency,
it will now kill itself (without dumping core) with the same signal
warnings can be silenced by the --no-warnings (-n) option
[1]    82894 abort      clj -A:dev
```

If I use `juxt/crux-core {:mvn/version "20.04-1.8.2-alpha"}` then it works fine, 1.8.3 and above do not work.

Happy to provide more details of my system if you let me know what you need.

**Comment by jarohen** [comment]
thanks for raising @armincerf - have you got any hs_err_*.log files dumped in that directory?

**Comment by armincerf** [comment]
So I did a bit more experimentation, turns out I was pulling in a few other dependencies and I didn't get the issue if crux was the only one, but I do get it if asciidoctorj is also a dependancy, so a deps.edn file like 
```
{:deps {juxt/crux-core {:mvn/version "20.05-1.8.3-alpha"}
        org.asciidoctor/asciidoctorj {:mvn/version "2.2.0"}}}
```
will cause the crash but just crux by itself won't.

No log files that I could find in the project directory I'm afraid.

**Comment by jarohen** [comment]
Hmm, also tried that here and looks fine. Assuming you've tried asciidoctor on its own?

**Comment by armincerf** [comment]
Yes no problems there.. Also I tried it on both my linux laptop and other OSX machine and couldn't reproduce, but not sure what's different about my MacBook that would make it crash on there. Any idea where I can look to find some form of logs or a stack trace?

**Comment by jarohen** [comment]
It's not an error message I've come across, I'm afraid - closest thing I've seen is when we cause exceptions in RocksDB's C code, but that's a SIGSEGV rather than a SIGABRT. 

If I've understood correctly (first time I've come across it), SIGABRT's caused by something within the application throwing a 'fatal error', so next steps would be to figure out what's throwing it?

**Comment by jkxyz** [comment]
I'm also getting a SIGABRT on the latest Crux release on macOS, except for me it's occurring when `juxt/crux-rocksdb` is on the classpath. 

I've tested with Java 11, 12, and 13, and I get the following with Java 11 and 13, but Java 12 works fine.

```
$ clj -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.10.1"} juxt/crux-core {:mvn/version "20.06-1.9.1-beta"} juxt/crux-rocksdb {:mvn/version "20.06-1.9.1-beta"}}}'
Clojure 1.10.1
user=> (System/getProperty "java.version")
"13.0.2"
user=> (require 'crux.api)
nil
user=> (def node (crux.api/start-node {:crux.node/topology ['crux.standalone/topology]}))
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
#'user/node
user=>
clojure: warning: clojure crashed, killed by SIGABRT.
```

No log files either. 

Please let me know if I can provide more info or try to debug more on my end. 

(Unfortunately on Java 12 I also get a JNA error from figwheel-main, so maybe there is just something peculiar about that release around FFI.)

**Comment by AndreasEdvardsson** [comment]
The solution for me was to use another JDK - just installing the Brew formulas `openjdk` or `openjdk@11` and attempting to start a node (with juxt/crux-core and juxt/crux-rocksdb on the classpath) resulted in a crash (interrupted by signal 6: SIGABRT).

By instead installing and using AdoptOpenJDK 11 I can successfully start a crux node.

**Comment by bbss** [comment]
I also ran into this, it also happens with crux-lmdb. AdoptOpenJDK seems to not have this issue, thanks @AndreasEdvardsson

**Comment by kevinmershon** [comment]
Just had this same issue today on OpenJDK15 on MacOS Catalina on Intel chipset

**Comment by brainocean** [comment]
Encounter same issue, change JDK to *AdoptOpenJDK8* thru homebrew as work-around now.

**Comment by stathissideris** [comment]
Same here!

**Comment by pekeler** [comment]
As suggested by @refset on [https://clojurians-log.clojureverse.org/crux/2021-04-19 ](https://webcache.googleusercontent.com/search?q=cache:ODOMg1hluPQJ:https://clojurians-log.clojureverse.org/crux/2021-04-19), setting env vars to disable libgcrypt can be a workaround.
```
$ CRUX_DISABLE_LIBGCRYPT=true CRUX_ENABLE_BYTEUTILS_SHA1=true clj -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.10.3"} pro.juxt.crux/crux-core {:mvn/version "1.18.0"} pro.juxt.crux/crux-rocksdb {:mvn/version "1.18.0"}}}'
Clojure 1.10.3
user=> (System/getProperty "java.version")
"11.0.10"
user=> (require 'crux.api)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
nil
user=> (crux.api/start-node {})
#<CruxNode>
```

**Comment by jarohen** [comment]
Thanks @pekeler! :pray:

@armincerf et al: does this fix the issue for you?

**Comment by pekeler** [comment]
Could the underlying reason be that crux.hash.jnr is loading _any_ version of the crypto library instead of a specific one? See https://github.com/cl-plus-ssl/cl-plus-ssl/pull/115 for more info.

**Comment by pekeler** [comment]
> underlying reason be that crux.hash.jnr is loading any version of the crypto library

yes, looks like it:
```Clojure
(for [p ["java.vendor" "java.version" "os.name" "os.version"]]
  (System/getProperty p)) ;; Oracle Corporation, 11.0.10, Mac OS X, 11.5
(definterface OpenSSL (^void OpenSSL_add_all_digests []))
(.load (jnr.ffi.LibraryLoader/create OpenSSL) "libcrypto.44.dylib") ;; works
(.load (jnr.ffi.LibraryLoader/create OpenSSL) "crypto") ;; crashes
```
